scripts/tutorial-uploader/lib/api.ts

```diff
@@ -0,0 +1,260 @@
+      .map((item: any) => item.id);
+    return matchingIds;
+  }
+
+  /**
+   * Like getIds (plural) but expects at most one match.
+   * Throws if more than one match is found.
+   */
+  async getId<
+    CollectionName extends StringKeyOf<LearningApiSchema>,
+    FieldName extends StringKeyOf<
+      ElementType<LearningApiSchema[CollectionName]>
+    >,
+    FieldValue extends ElementType<
+      LearningApiSchema[CollectionName]
+    >[FieldName],
+  >(
+    collection: CollectionName,
+    field: FieldName,
+    value: FieldValue,
+  ): Promise<string | null> {
+    const ids = await this.getIds(collection, field, value);
+    if (ids.length === 0) {
+      return null;
+    }
+    if (ids.length === 1) {
+      return ids[0];
+    }
+    throw new Error(
+      `Found ${ids.length} items for getId('${collection}', '${field}', '${value}'. ` +
+        `Expected one or none.`,
+    );
+  }
+
+  /**
+   * Tutorials can have many translations, but we only use English at the moment.
+   */
+  async getEnglishTranslationId(tutorialId: string): Promise<number> {
+    // TODO: This assumes the only translation is english (currently true)
+    const response = await this.client.request(
+      readItem("tutorials", tutorialId, { fields: ["translations"] }),
+    );
+    if (!response.translations) {
+      throw new Error(`No translations for tutorial ${tutorialId}`);
+    }
+    return response.translations[0] as number;
+  }
+
+  /**
+   * Remove all topics from a tutorial.
+   * This works by deleting appropriate entries of
+   * "tutorials_tutorials_topics": a collection of one-to-many mappings of
+   * "tutorials" to "tutorials_topics".
+   */
+  async clearTopics(tutorialId: string) {
+    // "tutorials_tutorials_topics" is mapping of tutorial to topics
+    const ids = await this.getIds(
+      "tutorials_tutorials_topics",
+      "tutorials_id",
+      tutorialId,
+    );
+    for (const id of ids) {
+      await this.client
+        .request(deleteItem("tutorials_tutorials_topics", id))
+        .catch((err: any) => console.log(err));
+    }
+  }
+
+  async updateTutorialTopics(tutorialId: string, topicNames: string[]) {
+    await this.clearTopics(tutorialId);
+    for (const name of topicNames) {
+      const id = await this.getId("tutorials_topics", "name", name);
+      if (id === null) throw new Error(`No topic with name '${name}'`);
+      await this.client.request(
+        createItem("tutorials_tutorials_topics", {
+          tutorials_id: tutorialId,
+          tutorials_topics_id: id,
+        }),
+      );
+    }
+  }
+
+  /* Returns the file's ID */
+  async uploadLocalFolder(path: string): Promise<string> {
+    // Zip folder
+    const zippedFilePath = `${tmpdir()}/${randomBytes(8).toString(
+      "hex",
+    )}/tutorial.zip`;
+    await $`mkdir -p ${dirname(zippedFilePath)}`;
+    await $`(cd ${dirname(path)} && zip -qr ${zippedFilePath} ${basename(
+      path,
+    )})`;
```

Comment:
Probably more clear to use `cd` from zx: https://google.github.io/zx/api#cd

Should this be changing the cwd for all future processes, or only this specific one?

Reply:
I think that the intention is only to avoid having the full directory structure inside the zip file, but correct me if I'm wrong Frank. 

FWIW, for that, we could also modify the working directory with the `env -C` command in front of the `zip` command, something like `env -C ${path} zip -qr ${zippedFilePath} .` I like the use of `cd` as well, so that's just an alternative ([example](https://superuser.com/a/1842149))

Reply:
Yep, I don't think zx cd is ideal as I don't want to change the working directory for other commands. @Eric-Arellano @arnaucasau do you have a preference between `env -C` or `(cd ... && ... )`. I'll add an explanation comment either way.

Reply:
`(cd ... && ... )` sgtm

Reply:
Yeah, I agree with Eric :+1: